File: 3/part2.cpp
#include <iostream>
#include <fstream>
#include <vector>

using namespace std;


bool in(int elem, const vector<int>& data) {
    for (int e : data) {
        if (e == elem) return true;
    }
    return false;
}

int main() {
    int a = 0b100000110000;
    int b = 0b001110101101;
    cout << a << " " << b << endl;
    cout << a * b;

    string line;
    fstream File("../input.txt");

    vector<vector<int>> data = {};
    vector<int> ban = {};

    while (getline(File, line)) {
        vector<int> x = {};
        x.reserve(12);
        for (int i = 0; i < 12; i++) {
            x.push_back(line[i] - '0');
        }
        data.push_back(x);
    }

    int position = 0;
    while (data.size() != 1) {
        int common = 0;

        for (const vector<int> &i : data) {
            for (int j : i) {
                if ( !in(j, ban)) {
                    common += j - '0';
                }
            }
        }
        if (common < 500) common = 0;
        else common = 1;

        for (int i = 0; i < data.size(); i++) {
            if (data[i][position] != common) {
                ban.push_back(i);
            }
            cout << endl;
        }
        position++;

        cout << position << endl;
        for (auto & i : data) {
            for (int j : i) {
                cout << j << ' ';
            }
            cout << endl;
        }
        cout << "---------" << endl;
    }

    return 0;
}

File: 3/part2.py
with open('input.txt') as f:
    data = f.read().split()

position = 0
while len(data) != 1:
    common = 0

    for x in data:
        common += int(x[position])

    if (len(data) - common) > common:
        common = '0'
    else:
        common = '1'

    new_data = []
    for i in range(len(data)):
        if data[i][position] == common:
            new_data.append(data[i])

    position += 1

    data = list(new_data)
    new_data = []

    print(len(data), data)

o2 = int(data[0], 2)

with open('input.txt') as f:
    data = f.read().split()

position = 0
while len(data) != 1:
    common = 0

    for x in data:
        common += int(x[position])

    if (len(data) - common) > common:
        common = '0'
    else:
        common = '1'

    new_data = []
    for i in range(len(data)):
        if data[i][position] != common:
            new_data.append(data[i])

    position += 1

    data = list(new_data)
    new_data = []

    print(len(data), data)

co2 = int(data[0], 2)

print(o2, co2)
print(o2 * co2)
